File: resources/web/wwi/webotsjs/WbDirectionalLight.js
// Copyright 1996-2021 Cyberbotics Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {WbLight} from "./WbLight.js"

class WbDirectionalLight extends WbLight {
  constructor(id, on, color, direction, intensity, castShadows, ambientIntensity){
    super(id, on, color, intensity, castShadows, ambientIntensity);
    this.wrenLight = undefined;
    this.direction = direction;
  }

  delete(){
    if (this.wrenObjectsCreatedCalled)
      _wr_node_delete(this.wrenLight);

    super.delete();
  }

  createWrenObjects() {
    this.wrenLight = _wr_directional_light_new();
    super.createWrenObjects();

    this.applyLightDirectionToWren();
  }

  applyLightDirectionToWren() {
    let pointer = _wrjs_color_array(this.direction.x, this.direction.y, this.direction.z);
    _wr_directional_light_set_direction(this.wrenLight, pointer);
  }

  applyLightColorToWren() {
    let pointer = _wrjs_color_array(this.color.x, this.color.y, this.color.z);

    _wr_directional_light_set_color(this.wrenLight, pointer);
  }

  applyLightIntensityToWren() {
    _wr_directional_light_set_intensity(this.wrenLight, this.intensity);
  }

  applyLightVisibilityToWren() {
    _wr_directional_light_set_on(this.wrenLight, this.on);

    let maxCount = _wr_config_get_max_active_directional_light_count();
    let activeCount = _wr_scene_get_active_directional_light_count(_wr_scene_get_instance());
    if (activeCount === maxCount)
      console.log("Maximum number of directional lights " + maxCount +" has been reached, newly added lights won't be rendered.");
  }

  applyLightShadowsToWren() {
    _wr_directional_light_set_cast_shadows(this.wrenLight, this.castShadows);
  }

  clone(customID) {
    return new WbDirectionalLight(customID, this.on, this.color, this.direction, this.intensity, this.castShadows, this.ambientIntensity);
  }
}
export {WbDirectionalLight}

File: src/webots/nodes/WbDirectionalLight.cpp
// Copyright 1996-2023 Cyberbotics Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "WbDirectionalLight.hpp"
#include "WbField.hpp"
#include "WbSFBool.hpp"
#include "WbSFColor.hpp"
#include "WbSFDouble.hpp"
#include "WbSFVector3.hpp"
#include "WbWrenRenderingContext.hpp"

#include <wren/config.h>
#include <wren/directional_light.h>
#include <wren/node.h>
#include <wren/scene.h>

void WbDirectionalLight::init() {
  mWrenLight = NULL;
  mDirection = findSFVector3("direction");
}

WbDirectionalLight::WbDirectionalLight(WbTokenizer *tokenizer) : WbLight("DirectionalLight", tokenizer) {
  init();
  if (tokenizer == NULL)
    mDirection->setValueNoSignal(0, -1, 0);
}

WbDirectionalLight::WbDirectionalLight(const WbDirectionalLight &other) : WbLight(other) {
  init();
}

WbDirectionalLight::WbDirectionalLight(const WbNode &other) : WbLight(other) {
  init();
}

void WbDirectionalLight::preFinalize() {
  WbLight::preFinalize();

  updateDirection();
}

void WbDirectionalLight::postFinalize() {
  WbLight::postFinalize();

  connect(mDirection, &WbSFVector3::changed, this, &WbDirectionalLight::updateDirection);
}

WbDirectionalLight::~WbDirectionalLight() {
  if (areWrenObjectsInitialized())
    wr_node_delete(WR_NODE(mWrenLight));
}

void WbDirectionalLight::createWrenObjects() {
  mWrenLight = wr_directional_light_new();
  WbLight::createWrenObjects();  // sets properties common to all lights
  applyLightDirectionToWren();
}

void WbDirectionalLight::updateDirection() {
  if (areWrenObjectsInitialized())
    applyLightDirectionToWren();
  emit directionChanged();
}

void WbDirectionalLight::applyLightIntensityToWren() {
  wr_directional_light_set_intensity(mWrenLight, mIntensity->value());
}

void WbDirectionalLight::applyLightColorToWren() {
  const float color[] = {static_cast<float>(mColor->red()), static_cast<float>(mColor->green()),
                         static_cast<float>(mColor->blue())};
  wr_directional_light_set_color(mWrenLight, color);
}

void WbDirectionalLight::applyLightVisibilityToWren() {
  wr_directional_light_set_on(mWrenLight, mOn->value());

  const int maxCount = wr_config_get_max_active_directional_light_count();
  const int activeCount = wr_scene_get_active_directional_light_count(wr_scene_get_instance());
  if (activeCount == maxCount)
    parsingWarn(
      tr("Maximum number of directional lights (%1) has been reached, newly added lights won't be rendered.").arg(maxCount));
}

void WbDirectionalLight::applyLightShadowsToWren() {
  wr_directional_light_set_cast_shadows(mWrenLight, mCastShadows->value());
}

void WbDirectionalLight::applyLightDirectionToWren() {
  float d[] = {static_cast<float>(mDirection->x()), static_cast<float>(mDirection->y()), static_cast<float>(mDirection->z())};
  wr_directional_light_set_direction(mWrenLight, d);
}

const WbVector3 &WbDirectionalLight::direction() const {
  return mDirection->value();
}

QStringList WbDirectionalLight::fieldsToSynchronizeWithX3D() const {
  QStringList fields;
  fields << "direction" << WbLight::fieldsToSynchronizeWithX3D();

  return fields;
}
